### PR TITLE
Added logging to n8n

### PR DESCRIPTION
**Pull request by krynble**
This commit adds logging to n8n using the Winston library.

For now, this commit only allows logging to console (default behavior)
or file (need to pass in config via environment variables).

Other logging methods can be further implemented using hooks. These were
skipped for now as it would require adding more dependencies.

Logging level is `info`by default, meaning no additional messages would
be displayed at the moment. Logging level can be set to verbose or debug as
well to enrich the generated logs.

The ILogger interface was added to the workflow project as it would make
it available for all other projects but the implementation was done on
the cli project.

### CODE REVIEW

File: packages/cli/src/ActiveWorkflowRunner.ts
Comment (janober) [inline packages/cli/src/ActiveWorkflowRunner.ts:62]: Can we not do that 1x outside of the class so that we do not have this line multiple times in the same file and can simply use directly `logger` everywhere inside of it?

File: packages/workflow/src/Interfaces.ts
Comment (ivov) [inline packages/workflow/src/Interfaces.ts:763]: Or `LogTypes` since this is a union type, not an interface.

**Comment by krynble** [comment]
Thank you for your contributions @MedAliMarz 

I had to revert back some of the messages from the logger to console as some information is only relevant to the terminal output and have little / no relevance to be sent to a logger system (files, elasticsearch, loggly, etc.).

Also I added shortcuts for the `webhook` and `worker` commands @janober Let me know if you see any issues with this.

File: packages/cli/commands/execute.ts
Comment (mutdmour) [inline packages/cli/commands/execute.ts:175]: nitpick - typo. 
```suggestion
				console.info('Execution was NOT successful. See log message for details.');
```

File: packages/cli/commands/export/credentials.ts
Comment (mutdmour) [inline packages/cli/commands/export/credentials.ts:17]: it looks like "logOutput" function from this file is not used anymore so maybe we should delete that

File: packages/cli/commands/import/credentials.ts
Comment (ahsanv) [inline packages/cli/commands/import/credentials.ts:60]: instead of using `console.log`, why don't we use Logger for all logs, and setup transports according to log level? Say info and lower (more important) go to file and standard output, and higher than info (debug etc) go to only file?

So that we have a consistent implementation?

File: packages/cli/commands/execute.ts
Comment (mutdmour) [inline packages/cli/commands/execute.ts:186]: nitpick - typo
```suggestion
			console.info('Execution was successful:');
```

File: packages/cli/src/Logger.ts
Comment (mutdmour) [inline packages/cli/src/Logger.ts:34]: should not we add some form of log rotation? or is that handled elsewhere?
file transport format supports maxsize and maxFiles to support this https://github.com/winstonjs/winston/blob/master/docs/transports.md#file-transport

File: packages/cli/commands/import/credentials.ts
Comment (krynble) [reply packages/cli/commands/import/credentials.ts:60]: Hey Ahsan.

This was a dillema during development. In some cases, I only want messages to be output to terminal only (they don't really make sens to be added to a log file as this is an import credentials command).

So I was using both the console.* messages with LoggerProxy.* messages according to the type of message.

For a log file, my expectations as a developer is to always have all errors and even better, have debug messages on log files to help me solve issues. So actually log files should contain all relevant logs and UI prompts (even if the interface is the terminal) should not be there.

Does this make sense? Let me know your thoughts!

File: packages/cli/commands/execute.ts
Comment (krynble) [reply packages/cli/commands/execute.ts:175]: Thanks, will merge all suggestions into a single commit =)

File: packages/cli/commands/export/credentials.ts
Comment (krynble) [reply packages/cli/commands/export/credentials.ts:17]: Thanks, removed!

File: packages/cli/src/Logger.ts
Comment (krynble) [reply packages/cli/src/Logger.ts:34]: Hi @mutdmour 

I am being very conservative to adding dependencies and it seems like adding log rotations would require another one.

The thoughts behind this first implementation is to create a basic logging framework that can be improved later, so for this reason I did not add any dependency for log output as all required new dependencies. Transports like elastic, loggly and cloudwatch logs all require more dependencies.

Maybe we can allow users to plug in other transports as they need. What do you think?

Comment (mutdmour) [reply packages/cli/src/Logger.ts:34]: adding maxsize and maxFiles should not require any new dependencies. File transport used here already supports these https://github.com/winstonjs/winston/blob/master/docs/transports.md#file-transport if I am not mistaken 
I have seen winston log files fill up disk before in a webapp without `maxsize` being set.

Comment (krynble) [reply packages/cli/src/Logger.ts:34]: Thanks for clarifying that @mutdmour .

I just realized that daily rotation would require a new lib but adding these settings do not.

Will apply those changes!

File: packages/cli/src/Logger.ts
Comment (mutdmour) [inline packages/cli/src/Logger.ts:36]: thanks for addressing this
one more thought - there might be best as configurable values rather than static

Comment (krynble) [reply packages/cli/src/Logger.ts]: Thanks again for your contribution @mutdmour 

Just updated with your suggestions!

File: packages/cli/config/index.ts
Comment (mutdmour) [inline packages/cli/config/index.ts:601]: nitpick - format here can be Number

**Review by mutdmour — APPROVED** [review]
LGTM - just a small nitpick

File: packages/cli/config/index.ts
Comment (krynble) [reply packages/cli/config/index.ts]: Thanks again!

File: packages/cli/config/index.ts
Comment (mutdmour) [inline packages/cli/config/index.ts]: default values need to be updated too since type is now Number

Comment (krynble) [reply packages/cli/config/index.ts]: Again: thanks, again :D hahaha

Comment (mutdmour) [reply packages/cli/config/index.ts]: 😊✅

**Review by ahsanv — APPROVED** [review]
👍 Great work!

File: packages/cli/src/Logger.ts
Comment (ivov) [inline packages/cli/src/Logger.ts:9]: To keep this as a singleton:
```suggestion
class Logger implements ILogger {
```

File: packages/workflow/src/LoggerProxy.ts
Comment (ivov) [inline packages/workflow/src/LoggerProxy.ts:13]: To make this more explicit than just "instance":

```suggestion
export function getLogger(): ILogger {
```

File: packages/workflow/src/LoggerProxy.ts
Comment (ivov) [inline packages/workflow/src/LoggerProxy.ts:10]: To keep consistency with the getters:

```suggestion
let logger: ILogger | undefined;

export function init(loggerInstance: ILogger) {
	logger = loggerInstance;
```

File: packages/cli/src/ActiveWorkflowRunner.ts
Comment (ivov) [inline packages/cli/src/ActiveWorkflowRunner.ts:68]: When do we call the logger as `LoggerProxy.getInstance();` (see above) and when as `getLogger` (originally `getInstance`)? The interfaces are identical.

File: packages/cli/src/ActiveWorkflowRunner.ts
Comment (ivov) [inline packages/cli/src/ActiveWorkflowRunner.ts:99]: Why not `LoggerProxy.verbose()` instead? Same for other examples in this file.

File: packages/cli/src/WebhookHelpers.ts
Comment (ivov) [inline packages/cli/src/WebhookHelpers.ts:33]: Maybe we do not need as much detail in the name when the logger is being used.

```suggestion
	LoggerProxy as Logger,
```

File: packages/cli/src/WorkflowExecuteAdditionalData.ts
Comment (ivov) [inline packages/cli/src/WorkflowExecuteAdditionalData.ts:39]: For readability:

```suggestion
	LoggerProxy as Logger,
```

File: packages/cli/src/WorkflowRunner.ts
Comment (ivov) [inline packages/cli/src/WorkflowRunner.ts:32]: Unused imports:

```suggestion
	IExecutionError,
```

File: packages/cli/src/WorkflowRunner.ts
Comment (ivov) [inline packages/cli/src/WorkflowRunner.ts:34]:
```suggestion
	LoggerProxy as Logger,
```

File: packages/core/src/ActiveWorkflows.ts
Comment (ivov) [inline packages/core/src/ActiveWorkflows.ts:10]:
```suggestion
	LoggerProxy as Logger,
```

File: packages/workflow/src/LoggerProxy.ts
Comment (ivov) [inline packages/workflow/src/LoggerProxy.ts:28]: Why not call it like this? Same for the other convenience methods.

```suggestion
	getLogger().debug(message, meta);
```

**Review by ivov — CHANGES_REQUESTED** [review]
Omar, great job! We had a number of comments regarding importing and usage of the `Logger` vs. `LoggerProxy`, how to use them and the reason behind the convenience methods and when they are not used. Happy to jump on a quick call with you. 

Looking forward for the feature.

Done together with @BHesseldieck.

File: packages/cli/src/ActiveWorkflowRunner.ts
Comment (krynble) [reply packages/cli/src/ActiveWorkflowRunner.ts:68]: Hello @ivov 

You should usually only use `LoggerProxy.getInstance()` as it should be available project-wide.

The `getLogger` function is used to initialize the logger for the first time and it is then provided to the `LoggerProxy`.

In a few cases, when the LoggerProxy has not been initialized yet you can use `getLogger` directly but it is recommended avoiding it as much as possible.

If it's unclear please let me know and I can clarify.

File: packages/workflow/src/LoggerProxy.ts
Comment (krynble) [reply packages/workflow/src/LoggerProxy.ts:28]: I like the idea of calling the convenience methods statically. Changed the code so it looks like this:

```
import {
    LoggerProxy as Logger,
}  from 'n8n-workflow';

...

LoggerProxy.log(...);
LoggerProxy.error(...);
```

**Comment by krynble** [comment]
Thanks a lot @ivov @ahsanv @mutdmour and @BHesseldieck  for your contributions!

May I ask for one last look before we merge? I would also ask you to approve changes if it's all good =)

File: packages/cli/commands/execute.ts
Comment (ivov) [inline packages/cli/commands/execute.ts:176]: Here as well, when do we use `console` and when `logger`?

File: packages/cli/commands/webhook.ts
Comment (ivov) [inline packages/cli/commands/webhook.ts:52]: Can we do it like this? (whole file)

```suggestion
		LoggerProxy.info(`\nStopping n8n...`);
```

File: packages/cli/commands/worker.ts
Comment (ivov) [inline packages/cli/commands/worker.ts:82]: Same as prior comment.

```suggestion
		LoggerProxy.info(`\nStopping n8n...`);
```

File: packages/cli/src/ActiveWorkflowRunner.ts
Comment (ivov) [inline packages/cli/src/ActiveWorkflowRunner.ts:67]: Maybe it would be more appropriate for this to be `console.info()`?

File: packages/cli/src/Logger.ts
Comment (ivov) [inline packages/cli/src/Logger.ts:16]: Maybe:

```suggestion
		const logFormat = winston.format.printf(({ message }) => message) as winston.Logform.Format;
```

**Review by ivov — APPROVED** [review]
Thanks Omar! Some minor remarks and suggestions.

File: packages/cli/commands/worker.ts
Comment (mutdmour) [inline packages/cli/commands/worker.ts:82]: is this `\n` needed?

File: packages/cli/src/ActiveWorkflowRunner.ts
Comment (mutdmour) [inline packages/cli/src/ActiveWorkflowRunner.ts:65]: is this `\n` needed?

File: packages/cli/src/WorkflowExecuteAdditionalData.ts
Comment (mutdmour) [inline packages/cli/src/WorkflowExecuteAdditionalData.ts:135]: why not include workflowId in logs in this file too? as well as sessionId? would make debugging that much easier.

File: packages/cli/src/Logger.ts
Comment (mutdmour) [inline packages/cli/src/Logger.ts:28]: nitpick - spacing here seems off

Comment (krynble) [reply packages/cli/src/Logger.ts:28]: Hi @mutdmour 

I did not find any issue here, spacing seems fine (once before and after each pipe). Any suggestions?

Comment (mutdmour) [reply packages/cli/src/Logger.ts:28]: my bad. probably just looked funky on github

File: packages/cli/src/Logger.ts
Comment (mutdmour) [inline packages/cli/src/Logger.ts:70]: 😊

**Comment by janober** [comment]
Thanks a lot. Got merged.

**Comment by janober** [comment]
Got released with n8n@0.118.0

**Pull request merged**
